src/App/Fossa/VSI/DynLinked/Internal/Resolve.hs

```diff
@@ -0,0 +1,86 @@
+toSourceUnit root dependencies = do
+  let (resolved, unresolved) = sortResolvedUnresolved dependencies
+  binaries <- traverse (analyzeSingleBinary root) unresolved
+
+  let project = toProject root $ Graphing.fromList (map toDependency resolved)
+  let unit = Srclib.toSourceUnit False project
+  pure $ unit{additionalData = fmap toDepData (Just binaries)}
+  where
+    toDepData d = AdditionalDepData (Just d) Nothing
+    toProject dir graph = ProjectResult VsiProjectType dir graph Complete []
+
+toDependency :: ResolvedLinuxPackage -> Dependency
+toDependency ResolvedLinuxPackage{..} = case resolvedLinuxPackageManager of
+  LinuxPackageManagerDEB -> renderDEB resolvedLinuxPackageMetadata
+  LinuxPackageManagerRPM -> renderRPM resolvedLinuxPackageMetadata
+  LinuxPackageManagerAPK -> renderAPK resolvedLinuxPackageMetadata
+  where
+    render fetcher projectParts revisionParts =
+      Dependency
+        { dependencyType = fetcher
+        , dependencyName = fromParts projectParts
+        , dependencyVersion = Just . CEq $ fromParts revisionParts
+        , dependencyLocations = []
+        , dependencyEnvironments = mempty
+        , dependencyTags = mempty
+        }
+    fromParts = intercalate "#"
+    renderDEB LinuxPackageMetadata{..} =
+      render
+        LinuxDEB
+        [linuxPackageID, linuxPackageDistro, linuxPackageDistroRelease]
+        [linuxPackageArch, linuxPackageRevision]
+    renderRPM LinuxPackageMetadata{..} =
+      render
+        LinuxRPM
+        [linuxPackageID, linuxPackageDistro, linuxPackageDistroRelease]
+        [linuxPackageArch, epoch <> linuxPackageRevision]
+    renderAPK LinuxPackageMetadata{..} =
+      render
+        LinuxAPK
+        [linuxPackageID, linuxPackageDistro, linuxPackageDistroRelease]
+        [linuxPackageArch, linuxPackageRevision]
+    epoch = maybe "" (<> ":") $ linuxPackageDistroEpoch resolvedLinuxPackageMetadata
+
+-- | Sort @Set DynamicDependency@ into two lists: "resolved", and "unresolved", in a single pass.
+--
+-- Whether a dependency is "resolved" is predicated on whether its @dynamicDependencyResolved@ is not @Nothing@.
+-- Dependencies thus sorted are unwrapped into more specific types based on the needs of downstream functions
+-- converting this set of dependencies into a @SourceUnit@.
+--
+-- This function reverses the order of dependencies for performance reasons,
+-- but dependency order isn't significant so we don't reverse it back.
+sortResolvedUnresolved :: Set DynamicDependency -> ([ResolvedLinuxPackage], [Path Abs File])
+sortResolvedUnresolved = buckets [] [] . toList
+  where
+    buckets resolved unresolved [] = (resolved, unresolved)
+    buckets resolved unresolved (dep : remaining) = case dynamicDependencyResolved dep of
+      Nothing -> buckets resolved (dynamicDependencyDiskPath dep : unresolved) remaining
+      Just linuxPackage -> buckets (linuxPackage : resolved) unresolved remaining
```

Comment:
[optional] in the same way that general `for` loops make me think too hard, general recursion also makes me think too hard, even for really simple snippets like this one

this case is really simple and perfectly fine as-is, so this is genuinely optional. here's an alternative implementation that might make it ever so slightly clearer / more-"obviously correct":

We can break this up into two smaller steps we can compose:
```hs
step1 :: [DynamicDependency] -> [Either ResolvedLinuxPackage (Path Abs File)]
step1 = undefined

step2 :: [Either ResolvedLinuxPackage (Path Abs File)] -> ([ResolvedLinuxPackage], [Path Abs File])
step2 = undefined

sortResolvedUnresolved :: Set DynamicDependency -> ([ResolvedLinuxPackage], [Path Abs File])
sortResolvedUnresolved = step2 . step1 . Set.toList
```

`step1` is fairly straightforward, where we'd map some function `DynamicDependency -> Either ResolvedLinuxPackage (Path Abs File)` into the list:

```hs
step1 :: [DynamicDependency] -> [Either ResolvedLinuxPackage (Path Abs File)]
step1 = map _something
```

`step2` is already defined for us in `Data.Either`:
```hs
partitionEithers :: [Either a b] -> ([a],[b])
```

..and putting it all together:

```hs
sortResolvedUnresolved :: Set DynamicDependency -> ([ResolvedLinuxPackage], [Path Abs File])
sortResolvedUnresolved = partitionEithers . map something . Set.toList
  where
    something :: DynamicDependency -> Either ResolvedLinuxPackage (Path Abs File)
    something dep = case ... of ...
```

---

There are a couple of reasons I say this is more "obviously-correct":

1. We know that `map`/`partitionEithers` will preserve the same number of elements, so we don't need to check our general-recursive version to make sure we're not dropping/adding elements. And their very-general types ensure they're not doing anything tricky with the provided values
2. We've narrowed the interesting logic into this `something` function with a much narrower type than our `buckets` from before. There are fewer things that can go wrong

---

Lastly, it's worth noting that this `partitionEithers . map something` implementation retains the same order of elements in the provided list.

`buckets` is essentially a hand-rolled `foldl` over the list, whereas `partitionEithers` uses a `foldr`. A useful exercise might be to implement with both `foldl` and `foldr` versions of something like `partitionEithers` -- possibly evaluating by hand -- to see where and why they differ

Reply:
This is such good feedback, thank you!